Allow a per-project custom character suite

cpeel opened the pull request:
Allow each project to include a set of characters that can be included as adhoc characters for the project. These appear in a special 'Proj' menu in the character picker.

This is available in the [per-project-charsuite](https://www.pgdp.org/~cpeel/c.branch/per-project-charsuite/) sandbox.

srjfoo commented:
I'm not entirely sure what's going on here, or if we need to worry about it at this point. This is a screenshot of some testing I did in the [Tiger Lily](https://www.pgdp.org/~cpeel/c.branch/per-project-charsuite/project.php?id=projectID5df912d3cae6b) project in P1:
<img width="168" alt="image" src="https://user-images.githubusercontent.com/36800239/87759145-65e8bd80-c7d3-11ea-9261-6e8157a15ee8.png">

I went through my OS's character viewer and picked some random Astral Plane charcters in addition to including a few of the zodiac symbols with VS-15. Also some plain ASCII characters.

1. WC flagging the characters as needing review is inconsistent.
2. WC flagging characters seems to depend on where they are in relation to the big filled-in circle.
3. I think the two characters on the far left for lines 2-4 are RTL -- when I was adding that character to the per-project character suite, the cursor kept re-positioning to the left when I tried to add another character to the right of it. Note that in the proofing interface, they just go where I put them. (Yes, I know we're not dealing with RTL right now -- I'm mentioning it only because of the positioning weirdness when I was entering that character in the project edit interface.)

srjfoo commented on tools/project_manager/editproject.php:
```diff
@@ -599,6 +603,40 @@ function set_from_post()
+                $errors .= _("The set of custom characters must be unique.")."<br>";
+            }
+
+            // all characters must be normalized
+            if(count(get_nonnormalized_codepoints($codepoints)))
+            {
+                $errors .= _("Custom characters must be in NFC normalized form.")."<br>";
+            }
```
Requiring characters to be normalized is legitimate, but for those entering combining diacritics, they may not even realize that what they're entering is not normalized. I deliberately added an `ā` both in normalized and non-normalized forms to see if that would trigger the message, and it did. Would it be realistic to normalize anything being added to a project custom character suite?

If all they're doing is copying a few characters from an existing character suite, or possibly from another web page, it should come to them normalized, but I'm thinking back to the few characters we found in project titles and authors' names that were non-normalized combining characters. It may be rare, but it will happen, and we'll likely wind up with confused PMs.

cpeel replied:
We can easily transparently normalize them and indeed that might be simpler for the users. My only worry is that some PM is going to put in a character that doesn't NFC normalize down to what they expect.

srjfoo replied:
Either they have to go out and find the normalized form, or they put in what they think is the right character (or character combination) and we normalize it for them. If it doesn't look right, they still have to go out and find the normalized form. As it stands now, it's slightly more work for them, because if they get the normalization error, they have to go out and find a normalized form when they might not have to if we do it for them.

I noticed one character yesterday where the character was vastly different, depending on the font, so it's also possible that any problems in shape might be due to font variations.

cpeel replied:
Characters are now normalized on save.

srjfoo replied:
Looks like the normalization is done after the check for duplicates:
<img width="121" alt="image" src="https://user-images.githubusercontent.com/36800239/88353774-67127100-cd24-11ea-891c-0b601a660b7b.png">

Reference custom charsuite in [UTF-8 practice: Tiger Lily and other stories.](https://www.pgdp.org/~cpeel/c.branch/per-project-charsuite/tools/charsuites.php?projectid=projectID5df912d3cae6b#projectID5df912d3cae6b).

cpeel replied:
Yes, it was on save to the DB. I've moved it up to before the validation. This is probably going to make it confusing for the user if they add a normalized letter and a non-normalized letter and then get a report that there is a duplicate. `<shrugs>`

srjfoo replied:
If you think it best, we can go back to just issuing the warning that characters in the custom suite have to be normalized, and not force the normalization. I'm trying to decide what I'd prefer as a PM, and I think I'd prefer the script to take care of it, especially if the only way I know to input it is non-normalized, but others might feel differently, and even if I get a duplicate character message.

Did the original version identify the non-normalized character(s)? (Short memory, here -- I don't remember!) If not, that's a potential problem if the PM doesn't have a clue what's normalized and what's not.

srjfoo commented on tools/project_manager/editproject.php:
```diff
@@ -599,6 +603,40 @@ function set_from_post()
+
+            // prevent disallowed characters from being added
+            $disallowed_codepoints = array_intersect(get_disallowed_codepoints(), $codepoints);
+            if($disallowed_codepoints != [])
```
Maybe add the space to the disallowed codepoints? This is what the picker tile looks like for a space: 
<img width="41" alt="image" src="https://user-images.githubusercontent.com/36800239/87844717-9a1cb680-c885-11ea-974d-58dee7974cf8.png">

cpeel replied:
We could special-case a space, but we can't add it to `get_disallowed_codepoints()` because it's allowed.

srjfoo replied:
Oops. Right. Silly me. I was focused on "disallowed for picker sets" not the fact that it would be disallowing the space completely. Yeah, bad idea.

cpeel commented:
> I did some more testing on WC and the special characters. As things are right now, they don't appear to be flagged as bad characters even if added as individual characters in both WC and the proofreading interface -- so if them not being flagged is going to be a problem we need to figure out how to deal with it (i.e. -- try to come up with a fix, or document?)/

Project custom characters are effectively treated the same way as those from character suites as far as WC is concerned. So while you may have found a bug, it is unlikely to be in the custom character code. Please do the same testing with those character in a character suite. I also don't understand how you expect them to be flagged -- did you add words with them to a project's bad list? Are you expecting them to be flagged because they are different scripts?

srjfoo commented:
> > I did some more testing on WC and the special characters. As things are right now, they don't appear to be flagged as bad characters even if added as individual characters in both WC and the proofreading interface -- so if them not being flagged is going to be a problem we need to figure out how to deal with it (i.e. -- try to come up with a fix, or document?)/
> 
> Project custom characters are effectively treated the same way as those from character suites as far as WC is concerned. So while you may have found a bug, it is unlikely to be in the custom character code. Please do the same testing with those character in a character suite. I also don't understand how you expect them to be flagged -- did you add words with them to a project's bad list? Are you expecting them to be flagged because they are different scripts?

I'll do some more research.

70ray commented:
Would it be better to specify the characters as their unicode value rather than pasting in the actual character? For (probably unrealistic) example the three characters 00D0 (capital eth), 0110 (D with stroke), and 0189 (African D) look pretty much identical so the user might not get what he really intends.

cpeel commented:
> Would it be better to specify the characters as their unicode value rather than pasting in the actual character? 

I thought about that, but then we have to teach them how to specify compound characters and also expect them to get the syntax right. Direct insert of the character seemed like the best UX.

srjfoo commented:
> > Would it be better to specify the characters as their unicode value rather than pasting in the actual character?
> 
> I thought about that, but then we have to teach them how to specify compound characters and also expect them to get the syntax right. Direct insert of the character seemed like the best UX.

I don't know about Linux, but both Macs and Windows machines have the ability to show a table of Unicode characters, I think? I know Macs do, but I'm not sure about the precise implementation on Windows). Taking another possible situation, hopefully someone actually going to the trouble of looking up characters in a table would be unlikely to confuse the ASCII letter `A` with its Greek or Cyrillic counterpart.

cpeel commented:
I added "N̅" to a project's custom character list (project was in English). I then put "N̅graze" into the proofreading interface and it did not get flagged -- which is expected since "N̅" is not a character in the English alphabet so aspell ignores the word (`<roll eyes>` thanks aspell). I added that word to the project's Bad Words list and WC correctly flagged it.

It's certainly possible we have some other WC-related issues in the code, but I don't think it's related to the per-project character suites

srjfoo commented:
> I added "N̅" to a project's custom character list (project was in English). I then put "N̅graze" into the proofreading interface and it did not get flagged -- which is expected since "N̅" is not a character in the English alphabet so aspell ignores the word (`<roll eyes>` thanks aspell). I added that word to the project's Bad Words list and WC correctly flagged it.
> 
> It's certainly possible we have some other WC-related issues in the code, but I don't think it's related to the per-project character suites

And I just now tried `Αño` (cap Greek alpha), and WC correctly identified as a possible bad word when I enabled the Spanish dictionary. On the other hand āño slid through without a whimper. I think that PM education is going to be very important. My earliest tests were aimed at seeing if WC would identify possible bad characters just on the basis of them not being part of a "real word".

srjfoo commented:
> > ... My earliest tests were aimed at seeing if WC would identify possible bad characters just on the basis of them not being part of a "real word".

Whoops ... posted too quickly. Those early tests can be discounted, I think, because I was trying a bunch of fairly bizarre stuff that should not happen in practice.
